### PR TITLE
feat: Added booking feature for rooms

### LINKED ISSUE #14: Booking: App (Bookings)
Create the UI and functionality to book on the App

### PR DESCRIPTION
**Pull request by Anders164a**
## Description

- Added booking feature for rooms

## Type of Change

- [x] ✨ New feature (non-breaking change which adds functionality)
- [ ] 🛠️ Bug fix (non-breaking change which fixes an issue)
- [ ] ❌ Breaking change (fix or feature that would cause existing functionality to change)
- [ ] 🧹 Code refactor
- [ ] ✅ Build configuration change
- [ ] 📝 Documentation
- [ ] 🗑️ Chore

### CODE REVIEW

File: lib/app/services/book.dart [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+        roomTypeIds.add(roomType.roomTypeId);
+      }
+
+      await _dio.post<Map<String, dynamic>>(
+        '/booking/rooms',
+        data: {
+          //🎖️
+          'start': '${start.toIso8601String()}Z',
+          'end': '${end.toIso8601String()}Z',
+          'roomTypeIds': roomTypeIds,
```
Comment (frederikpyt) [inline lib/app/services/book.dart]: 🥇

File: lib/book/view/hotel_room_book.dart [hunk 2]
```diff
@@ -0,0 +1,338 @@
+// ignore_for_file: use_build_context_synchronously
```
Comment (frederikpyt) [inline lib/book/view/hotel_room_book.dart]: ...

File: lib/book/service.dart [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+
+  final Dio _dio;
+
+  Future<List<Book>> getByHotelId(int hotelId) async {
+    final res = await _dio.get<List<dynamic>>('/hotel/$hotelId/room_types');
+
+    return res.data!.map<Book>((book) {
+      return Book.fromJSON(book as Map<String, dynamic>);
+    }).toList();
+  }
```
Comment (frederikpyt) [inline lib/book/service.dart]: ???

That endpoint returns roomtypes... We have no books in the backend

File: lib/book/model.dart [hunk 4]
```diff
@@ -0,0 +1,19 @@
+class Book {
```
Comment (frederikpyt) [inline lib/book/model.dart]: We have no books in the backend, what is this used for?!

File: lib/book/factory.dart [hunk 5]
```diff
@@ -0,0 +1,13 @@
+import 'package:overnites/app/factories/bearer_token_factory.dart';
+import 'package:overnites/app/factories/dio_factory.dart';
+import 'package:overnites/book/service.dart';
+
+class BookServiceFactory {
+  static Future<BookService> make() async {
+    return BookService(
```
Comment (frederikpyt) [inline lib/book/factory.dart]: Again, we have no books in this piece of software...

File: lib/book/view/hotel_room_book.dart [hunk 6]
```diff
@@ -0,0 +1,338 @@
...
+
+  @override
+  State<HotelBookPage> createState() => _HotelBookPageState();
+}
+
+class BookState {
+  BookState() {
+    _instance.roomTypeIds = [];
+    _instance.start = DateTime.now().subtract(const Duration(days: 1));
+    _instance.end = DateTime.now().subtract(const Duration(days: 1));
```
Comment (frederikpyt) [inline lib/book/view/hotel_room_book.dart]: Default to null?

File: lib/book/view/hotel_room_book.dart [hunk 7]
```diff
@@ -0,0 +1,338 @@
...
+  }
+  BookState._internal();
+
+  static final BookState _instance = BookState._internal();
+
+  List<BookRoomType> roomTypeIds = [];
+  DateTime start = DateTime.now().subtract(const Duration(days: 1));
+  DateTime end = DateTime.now().subtract(const Duration(days: 1));
```
Comment (frederikpyt) [inline lib/book/view/hotel_room_book.dart]: Default to null?

File: lib/book/view/hotel_room_book.dart [hunk 8]
```diff
@@ -0,0 +1,338 @@
...
+  List<Widget> selectedRooms = <Widget>[];
+
+  Future<void> getRoomTypesByHotelId() async {
+    final roomTypeService = await RoomTypeServiceFactory.make();
+
+    roomTypes = await roomTypeService.getByHotelId(widget.hotel.id);
+  }
+
+  final _bookState = BookState();
+  final GlobalKey key = GlobalKey();
```
Comment (frederikpyt) [inline lib/book/view/hotel_room_book.dart]: Where is this used?

File: lib/book/view/hotel_room_book.dart [hunk 9]
```diff
@@ -0,0 +1,338 @@
...
+  }
+  BookState._internal();
+
+  static final BookState _instance = BookState._internal();
+
+  List<BookRoomType> roomTypeIds = [];
+  DateTime start = DateTime.now().subtract(const Duration(days: 1));
+  DateTime end = DateTime.now().subtract(const Duration(days: 1));
+
+  void removeRoomTypeIfSet(ValueKey<dynamic> key) {
```
Comment (frederikpyt) [inline lib/book/view/hotel_room_book.dart]: Avoid dynamic type

File: lib/book/view/hotel_room_book.dart [hunk 10]
```diff
@@ -0,0 +1,338 @@
...
+    )) {
+      roomTypeIds.removeWhere(
+        (e) => e.roomTypeKey == key,
+      );
+    }
+  }
+}
+
+class BookRoomType {
+  BookRoomType(ValueKey<dynamic> key, int id) {
```
Comment (frederikpyt) [inline lib/book/view/hotel_room_book.dart]: Avoid dynamic type

File: lib/book/view/hotel_room_book.dart [hunk 11]
```diff
@@ -0,0 +1,338 @@
...
+    roomTypeKey = key;
+    roomTypeId = id;
+  }
+
+  late ValueKey<dynamic> roomTypeKey;
```
Comment (frederikpyt) [inline lib/book/view/hotel_room_book.dart]: Avoid dynamic type

File: lib/book/view/hotel_room_book.dart [hunk 12]
```diff
@@ -0,0 +1,338 @@
...
+                                    DateTime.now().isAfter(_bookState.end)) {
+                                  setState(() {
+                                    errorMessage = 'You need to pick a date';
+                                  });
+                                  return;
+                                } else if (_bookState.roomTypeIds.isEmpty) {
+                                  setState(() {
+                                    errorMessage =
+                                        'You need to pick minimum 1 room type '
+                                        "you'd like to book";
```
Comment (frederikpyt) [inline lib/book/view/hotel_room_book.dart]: l10n

File: lib/book/view/hotel_room_book.dart [hunk 13]
```diff
@@ -0,0 +1,338 @@
...
+                              setState(() {
+                                errorMessage = '';
+                              });
+
+                              if (await (await AuthServiceFactory.make())
+                                  .isAuthenticated()) {
+                                if (DateTime.now().isAfter(_bookState.start) ||
+                                    DateTime.now().isAfter(_bookState.end)) {
+                                  setState(() {
+                                    errorMessage = 'You need to pick a date';
```
Comment (frederikpyt) [inline lib/book/view/hotel_room_book.dart]: l10n

**Review by frederikpyt — CHANGES_REQUESTED** [review]

File: lib/book/model.dart (hunk 4 above)
Comment (Anders164a) [reply lib/book/model.dart]: Removed unused code

File: lib/book/service.dart (hunk 3 above)
Comment (Anders164a) [reply lib/book/service.dart]: Removed unused code

File: lib/book/factory.dart (hunk 5 above)
Comment (Anders164a) [reply lib/book/factory.dart]: Removed unused code

File: lib/book/view/hotel_room_book.dart (hunk 8 above)
Comment (Anders164a) [reply lib/book/view/hotel_room_book.dart]: Removed unused code

File: lib/book/view/hotel_room_book.dart (hunk 9 above)
Comment (Anders164a) [reply lib/book/view/hotel_room_book.dart]: Changes added

File: lib/book/view/hotel_room_book.dart (hunk 10 above)
Comment (Anders164a) [reply lib/book/view/hotel_room_book.dart]: Changes added

File: lib/book/view/hotel_room_book.dart (hunk 11 above)
Comment (Anders164a) [reply lib/book/view/hotel_room_book.dart]: Changes added

File: lib/book/view/hotel_room_book.dart (hunk 6 above)
Comment (Anders164a) [reply lib/book/view/hotel_room_book.dart]: Changes added

File: lib/book/view/hotel_room_book.dart (hunk 7 above)
Comment (Anders164a) [reply lib/book/view/hotel_room_book.dart]: Changes added

File: lib/book/view/hotel_room_book.dart (hunk 12 above)
Comment (Anders164a) [reply lib/book/view/hotel_room_book.dart]: no

File: lib/book/view/hotel_room_book.dart (hunk 13 above)
Comment (Anders164a) [reply lib/book/view/hotel_room_book.dart]: no

File: lib/app/services/book.dart (hunk 1 above)
Comment (Anders164a) [reply lib/app/services/book.dart]: image

File: lib/book/view/hotel_room_book.dart (hunk 2 above)
Comment (Anders164a) [reply lib/book/view/hotel_room_book.dart]: mhmm

File: lib/booking/view/hotel_room_booking.dart [hunk 14]
```diff
@@ -0,0 +1,337 @@
...
+                                if (await (await BookingServiceFactory.make())
+                                    .book(
+                                  _bookState.start!,
+                                  _bookState.end!,
+                                  _bookState.roomTypeIds,
+                                )) {
+                                  navigator.pop();
+                                }
+                              } else {
+                                // ignore: use_build_context_synchronously
```
Comment (frederikpyt) [inline lib/booking/view/hotel_room_booking.dart:288]: ;(

**Review by frederikpyt — APPROVED** [review]

File: lib/booking/view/hotel_room_booking.dart (hunk 14 above)
Comment (Anders164a) [reply lib/booking/view/hotel_room_booking.dart:288]: ;)

**Pull request merged**